src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -601,28 +601,20 @@ internal MemberAccessor MemberAccessorStrategy
 
         /// <summary>
         /// Initializes the converters for the reflection-based serializer.
-        /// <seealso cref="InitializeForReflectionSerializer"/> must be checked before calling.
         /// </summary>
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        internal void InitializeForReflectionSerializer()
+        [MemberNotNull(nameof(s_reflectionTypeInfoResolver))]
+        internal static void EnsureInitializedForReflectionSerializer()
```

Comment:
Why the change to `static`?

This was initially changed to a static  in https://github.com/dotnet/runtime/pull/64646 and changed back to an instance method in https://github.com/dotnet/runtime/pull/68496 due to issues that escape me right now. cc @eiriktsarpalis

Reply:
My changes unintentionally regressed .NET 6 behavior, namely that `JsonSerializerOptions` instances (including ones emanating from `JsonSerializerContext`) are individually and silently opted into reflection metadata generation when passed into one of the reflection-based serialization methods. Preserving this behaviour necessitates keeping a sentinel value on every options instance.

I'm honestly not a fan of the current behaviour, but if do decide to change it we should be deliberate about it and 1) mark it as a breaking change, 2) expose a knob in `JsonSerializerContext` so that users can easily get back the reflection fallback behavior.

Reply:
I understand the current behavior but why do we do that in the first place? Changing from throwing to not throwing is not a breaking change as far as I understand

Reply:
let's discuss this during our next feature crew meeting. I'm in favor of making breaking change here (i.e. disallow fallback always)

Reply:
> I understand the current behavior but why do we do that in the first place? Changing from throwing to not throwing is not a breaking change as far as I understand

It doesn't throw because it falls back to the reflection-derived metadata. This might be acceptable for CoreCLR applications, but it invalidates the purpose of source generators.